syncer/syncer.go

```diff
@@ -863,25 +866,37 @@ func (s *Syncer) updateReplicationLag(job *job, lagKey string) {
 				time.Sleep(time.Second * time.Duration(t))
 			}
 		})
-
-		switch job.tp {
-		case ddl, skip:
-			// NOTE: we handle ddl/skip job separately because ddl job will clean all the dml job before execution
-			lag = s.calcReplicationLag(int64(job.eventHeader.Timestamp))
-		default: // dml job
-			// NOTE workerlagmap already init all dml key(queueBucketName) before syncer running.
-			s.workerLagMap[lagKey].Store(s.calcReplicationLag(int64(job.eventHeader.Timestamp)))
-		}
+		lag = s.calcReplicationLag(int64(job.eventHeader.Timestamp))
+		s.workerLagArray[lagIdx].Store(lag)
 	}
+}
+
+func (s *Syncer) updateReplicationLagMetric() {
+	var lag int64
 	// find all job queue lag choose the max one
-	for _, l := range s.workerLagMap {
-		if wl := l.Load(); wl > lag {
+	for idx := range s.workerLagArray {
+		if wl := s.workerLagArray[idx].Load(); wl > lag {
 			lag = wl
 		}
 	}
 	metrics.ReplicationLagHistogram.WithLabelValues(s.cfg.Name, s.cfg.SourceID, s.cfg.WorkerName).Observe(float64(lag))
 	metrics.ReplicationLagGauge.WithLabelValues(s.cfg.Name, s.cfg.SourceID, s.cfg.WorkerName).Set(float64(lag))
 	s.secondsBehindMaster.Store(lag)
+
+	failpoint.Inject("ShowLagInLog", func(v failpoint.Value) {
+		minLag := v.(int)
+		if int(lag) >= minLag {
+			s.tctx.L().Info("ShowLagInLog", zap.Int64("lag", lag))
+		}
+	})
+
+	// reset ddl / skip lag in case of ddl / skip lag is never updated
+	if lag == s.workerLagArray[ddlLagIdx].Load() {
```

Comment:
This reset seems not reasonable. If a ddl need long time to execute and dml/skip have 0 latency. Then after the first 100ms,the ddl lag is zeroed, so after that the overall latency is 0 which is not reflect the truth. 

I suggest the ddl lag stores the ts that ddl job started and reset it to 0 when ddl job is finished. So we can calculate the true latency when execute a slow ddl event.

Reply:
It's a good idea and we can get one step closer
since there will also be some DML statements that execute very slowly, we can standardise the lag calculation logic for both DDL and DML , in https://github.com/pingcap/dm/pull/2060/commits/deaf4b47712631044d0191a5c680777bde05d144